\newpage

\subsection{Neuroevolution of Augmenting Topologies}
Neuroevolution uses evolutionary algorithms, such as genetic algorithms, to optimise ANNs \cite{neuroevolution_review}.
One such algorithm is Neuroevolution of Augmenting Topologies (NEAT), which evolves the topology and weights of ANNs \cite{neat_main, neat_short, neat_phd}.
It gained traction when published since it adressed some previous problems in neuroevolution and illustrated how the learning speed could
be increased by evolving the structure of the networks. Additional versions of NEAT has been created which enables evolving networks
of larger scale and complexity more efficiently.

NEAT uses a direct encoding; genomes contain connection and node genes which specify the network. Each connection gene is identified
by a number and references the two node genes it spans (see Figure \ref{mapping}). The node gene is also identified by a number.

\input{resources/tex/genotype_phenotype.tex}

The initial population of genomes consists of structurally identical and fully connected networks without any hidden nodes.
New structures emerge by mutations which can add new node or connection genes to the genomes. Nodes are added by splitting an existing
connection, which is disabled and two new connection genes are added along with the new node gene (see Figure \ref{node_mutation}).

\input{resources/tex/added_node_mutation.tex}

Mutations can also add new links, which connect any previously unconnected nodes (see Figure \ref{link_mutation}).
Self loops and recurrent connections are allowed.

The weights of the connections are also be mutated by being slightly perturbed. A list of additional mutations employed by NEAT are outlined in Table X.

\input{resources/tex/added_link_mutation.tex}

New structures can also be created during crossover, in which the genes of two genomes are combined.
The id attribute of the genes, referred to as an innovation number, serves to uniquely identify each mutation that has occured.
Each time a new link or node is added, a database containing all previous mutations and their ids is referenced.
As such, identical mutations are labeled with the same id.
These numbers are used during speciation and crossover. NEAT divides the population of genomes into species based on their structural
similarity, which is measured as a compatibility score.


\todo{Explain speciation and fitness sharing}

\todo{Add SUPER illustration of NEAT here}

\input{resources/tex/neat_overview.tex}